### ISSUE TITLE
Scylla does not reclaim disk space when keys expire via TTL

### ISSUE THREAD

**Issue by bancron** [scylladb/scylla#5168]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

Hello,

Moreno [suggested](https://scylladb-users.slack.com/archives/C2NLNBXLN/p1570720711032900?thread_ts=1570644344.008900&cid=C2NLNBXLN) I open this issue.

report_uuid=d4c5d9b5-f9a4-4ab2-abb4-221935185aaa

We have a Scylla cluster running on AWS. We are storing data with a TTL of 1 day (and a smaller amount of data with a TTL of 7 days).

Problem: the data does not seem to be removed from the disk upon expiry. We are using TimeWindowCompactionStrategy.

Since we are never deleting, only expiring via TTL, `gc_grace_seconds` is set to 0.

The very strange thing to me is that running `nodetool cleanup` will reclaim a very large portion of the used disk space. My understanding is that `nodetool cleanup` is only used when adding or removing nodes from the cluster. This cluster has always been the same size, with no nodes added or removed.

I am not sure what `nodetool cleanup` is doing. Should I set up a cron job to run it, since it seems to help?

Without running it, the cluster will eventually fill up completely and then die. This has already happened once, and we have deleted the cluster entirely and provisioned a new one.

We are not running prometheus, but here is a grafana graph of disk usage on /mnt (which is where we are storing the data):
![image](https://user-images.githubusercontent.com/3839100/66684956-15cf4000-ec30-11e9-8e52-23ddef40e0a2.png)

You can see the disk usage drops when I run `nodetool cleanup`.

*Installation details*
Scylla version (or git commit hash): 3.0.10
Cluster size: 10
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu

*Hardware details (for performance issues)*
Platform (physical/VM/cloud instance type/docker): AWS `i3en.6xlarge`
Hardware: sockets= cores=24 vCPU hyperthreading= memory=192 GiB
Disks: (SSD/HDD, count): 2 * 7500 GiB NVMe SSD

Schema: 
```
CREATE KEYSPACE tribute WITH replication = {'class': 'NetworkTopologyStrategy', 'us-east': '2'}  AND durable_writes = true;

CREATE TABLE tribute.impressions (
    k blob PRIMARY KEY,
    v blob,
    version int
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '1', 'compaction_window_unit': 'MINUTES'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 0
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

**Comment by bhalevy** [comment]
@bancron `nodetool cleanup` is running cleanup compaction and therefore it will delete expired data.
It is possible that you hit #2687 that is now fixed in master (but not yet ported to 3.0).

Can you please provide logs from compaction while setting both `compaction` and `compaction_manager` to trace so we can verify if this could be the case?

```
nodetool setlogginglevel compaction trace
nodetool setlogginglevel compaction_manager trace
```

Cc @raphaelsc @denesb

**Comment by bhalevy** [comment]
Ref #4969

**Comment by bancron** [comment]
Hi,

I have uploaded the logs under the uuid 44B8D417-209A-404F-BDD2-3DE425B8C056.

Are these the logs you wanted? Please LMK if there are any other logs I can provide.

**Comment by bancron** [comment]
Hi,

I am willing to build a scylla deb package from source to fix this issue. If I patch https://github.com/scylladb/scylla/commit/dd769431259ddfea6b7523e45635e643fa11ef58 into `scylla-3.1.0`, will that fix this issue?

**Comment by slivne** [comment]
@bancron please have a look at the specific instructions that helped in another case https://github.com/scylladb/scylla/issues/4969#issuecomment-530754258 (benny did reference this issie)

**Comment by bancron** [comment]
Hi @slivne, yes, I saw the link, thanks. It looks like the user has a lot of tombstoned data. Since we have no deletes in our usage and gc_grace_seconds = 0, I think this does not apply. I have run sstablemetadata and the Estimated droppable tombstones = 0.0.

**Comment by raphaelsc** [comment]
On Fri, Oct 18, 2019 at 5:03 PM Naomi B <notifications@github.com> wrote:

> Hi @slivne <https://github.com/slivne>, yes, I saw the link, thanks. It
> looks like the user has a lot of tombstoned data. Since we have no deletes
> in our usage and gc_grace_seconds = 0, I think this does not apply. I have
> run sstablemetadata and the Estimated droppable tombstones = 0.0.
>

For all sstables in the table? Estimated droppable tombstones = 0.0 means
that sstable has no droppable data at all. BTW, TTLd data is also tracked
in this estimation. Please share lines that follow 'Estimated tombstone
drop times:', from sstablemetadata output, if any.

**Comment by bancron** [comment]
Here is the output; I ran this on the most recent big-Data file.
```
SSTable: /mnt/scylla/data/tribute/impressions-275f4ee0e55111e99d61000000000005/la-84926-big
Partitioner: org.apache.cassandra.dht.Murmur3Partitioner
Bloom Filter FP chance: 0.010000
Minimum timestamp: 1571427553487422
Maximum timestamp: 1571428368310015
SSTable min local deletion time: 2147483647
SSTable max local deletion time: 1572033168
Compressor: org.apache.cassandra.io.compress.LZ4Compressor
Compression ratio: 0.8132619978660564
TTL min: 0
TTL max: 2147483647
Estimated droppable tombstones: 0.0
SSTable Level: 0
Repaired at: 0
Replay positions covered: {CommitLogPosition(segmentId=-1, position=0)=CommitLogPosition(segmentId=252201579136370485, position=12398600)}
totalColumnsSet: -1
totalRows: -1
Estimated tombstone drop times:
1571513961:      3244
1571513978:      3234
1571513994:      3190
1571514010:      3086
1571514025:      2774
1571514041:      3210
1571514059:      3324
1571514077:      3410
1571514095:      3328
1571514112:      3066
1571514127:      2868
1571514142:      2772
1571514158:      3358
1571514175:      3152
1571514191:      2762
1571514206:      2822
1571514222:      2960
1571514238:      3088
1571514253:      2922
1571514269:      3058
1571514285:      3050
1571514300:      2680
1571514314:      2776
1571514330:      2850
1571514344:      2790
1571514360:      3156
1571514377:      3362
1571514394:      3206
1571514411:      2836
1571514426:      2722
1571514440:      2694
1571514454:      2540
1571514467:      2524
1571514481:      2690
1571514497:      3060
1571514513:      3186
1571514530:      3386
1571514548:      3188
1571514564:      3066
1571514580:      3036
1571514597:      3524
1571514616:      3428
1571514634:      3400
1571514651:      3148
1571514668:      3164
1571514686:      3100
1571514702:      3058
1571514718:      2834
1571514732:      2610
1571514746:      2484
1571514760:      2524
1572032360:      1006
1572032379:      1172
1572032398:      1130
1572032416:      1054
1572032435:      1038
1572032453:       996
1572032469:       880
1572032482:       768
1572032496:       760
1572032511:       914
1572032527:       954
1572032542:       832
1572032557:       944
1572032575:      1132
1572032594:      1098
1572032614:      1144
1572032634:      1014
1572032653:      1088
1572032671:      1106
1572032689:      1152
1572032705:       908
1572032721:       918
1572032736:       838
1572032750:       872
1572032766:       844
1572032782:       850
1572032796:       810
1572032812:       980
1572032831:      1168
1572032851:      1182
1572032871:      1102
1572032890:      1004
1572032908:      1018
1572032924:       936
1572032941:      1044
1572032958:       908
1572032971:       626
1572032985:       910
1572033002:      1044
1572033020:      1132
1572033038:      1056
1572033055:       984
1572033071:       792
1572033085:       814
1572033099:       812
1572033113:       912
1572033130:       968
1572033147:       886
1572033161:       616
Count               Row Size        Cell Count
1                          0                 0
2                          0                 0
3                          0            100408
4                          0                 0
5                          0                 0
6                          0                 0
7                          0                 0
8                          0                 0
10                         0                 0
12                         0                 0
14                         0                 0
17                         0                 0
20                         0                 0
24                         0                 0
29                         0                 0
35                         0                 0
42                         0                 0
50                         0                 0
60                         0                 0
72                         0                 0
86                         0                 0
103                        0                 0
124                        0                 0
149                        0                 0
179                        0                 0
215                        0                 0
258                        0                 0
310                        0                 0
372                        0                 0
446                        0                 0
535                        0                 0
642                        0                 0
770                        0                 0
924                        0                 0
1109                       0                 0
1331                       0                 0
1597                       0                 0
1916                       0                 0
2299                       0                 0
2759                       0                 0
3311                       0                 0
3973                       0                 0
4768                     593                 0
5722                   14121                 0
6866                   25866                 0
8239                   20542                 0
9887                   13821                 0
11864                   8684                 0
14237                   5119                 0
17084                   4627                 0
20501                   2897                 0
24601                   1639                 0
29521                    882                 0
35425                    324                 0
42510                    186                 0
51012                    411                 0
61214                    340                 0
73457                    342                 0
88148                     12                 0
105778                     1                 0
126934                     0                 0
152321                     1                 0
182785                     0                 0
219342                     0                 0
263210                     0                 0
315852                     0                 0
379022                     0                 0
454826                     0                 0
545791                     0                 0
654949                     0                 0
785939                     0                 0
943127                     0                 0
1131752                    0                 0
1358102                    0                 0
1629722                    0                 0
1955666                    0                 0
2346799                    0                 0
2816159                    0                 0
3379391                    0                 0
4055269                    0                 0
4866323                    0                 0
5839588                    0                 0
7007506                    0                 0
8409007                    0                 0
10090808                   0                 0
12108970                   0                 0
14530764                   0                 0
17436917                   0                 0
20924300                   0                 0
25109160                   0                 0
30130992                   0                 0
36157190                   0                 0
43388628                   0                 0
52066354                   0                 0
62479625                   0                 0
74975550                   0                 0
89970660                   0                 0
107964792                  0                 0
129557750                  0                 0
155469300                  0                 0
186563160                  0                 0
223875792                  0                 0
268650950                  0                 0
322381140                  0                 0
386857368                  0                 0
464228842                  0                 0
557074610                  0                 0
668489532                  0                 0
802187438                  0                 0
962624926                  0                 0
1155149911                 0                 0
1386179893                 0                 0
1663415872                 0                 0
1996099046                 0                 0
2395318855                 0                 0
2874382626                 0
3449259151                 0
4139110981                 0
4966933177                 0
5960319812                 0
7152383774                 0
8582860529                 0
10299432635                 0
12359319162                 0
14831182994                 0
17797419593                 0
21356903512                 0
25628284214                 0
30753941057                 0
36904729268                 0
44285675122                 0
53142810146                 0
63771372175                 0
76525646610                 0
91830775932                 0
110196931118                 0
132236317342                 0
158683580810                 0
190420296972                 0
228504356366                 0
274205227639                 0
329046273167                 0
394855527800                 0
473826633360                 0
568591960032                 0
682310352038                 0
818772422446                 0
982526906935                 0
1179032288322                 0
1414838745986                 0
Estimated cardinality: 77
```
</p>
</details>

I then ran it on a file from the middle of the range of big-Data files. This one does show tombstones.
```
SSTable: /mnt/scylla/data/tribute/impressions-275f4ee0e55111e99d61000000000005/la-74749-big
Partitioner: org.apache.cassandra.dht.Murmur3Partitioner
Bloom Filter FP chance: 0.010000
Minimum timestamp: 1571314856201033
Maximum timestamp: 1571315495432551
SSTable min local deletion time: 2147483647
SSTable max local deletion time: 1571920295
Compressor: org.apache.cassandra.io.compress.LZ4Compressor
Compression ratio: 0.8047042740599222
TTL min: 0
TTL max: 2147483647
Estimated droppable tombstones: 0.684064357962502
SSTable Level: 0
Repaired at: 0
Replay positions covered: {CommitLogPosition(segmentId=-1, position=0)=CommitLogPosition(segmentId=234187180626884617, position=16412680)}
totalColumnsSet: -1
totalRows: -1
Estimated tombstone drop times:
1571314862:      3897
1571314875:      3618
1571314887:      3126
1571314898:      3567
1571314910:      3789
1571314923:      3786
1571314935:      3876
1571314947:      3381
1571314957:      3423
1571314968:      3222
1571314980:      3657
1571314991:      3477
1571315002:      3222
1571315014:      3192
1571315025:      3579
1571315037:      3948
1571315051:      3927
1571315063:      4002
1571315077:      4095
1571315089:      3690
1571315100:      3378
1571315112:      3570
1571315124:      3969
1571315137:      4167
1571315151:      3933
1571315164:      3561
1571315175:      3132
1571315186:      3456
1571315198:      3708
1571315210:      3990
1571315224:      3954
1571315236:      3684
1571315248:      3630
1571315260:      3468
1571315273:      4203
1571315286:      3909
1571315299:      4104
1571315312:      3942
1571315325:      3819
1571315339:      4725
1571315354:      4392
1571315367:      3867
1571315381:      3933
1571315393:      3219
1571315404:      3921
1571315417:      4125
1571315430:      4287
1571315445:      4653
1571315453:         3
1571315460:      4815
1571315475:      4215
1571315488:      3765
1571919663:      1354
1571919677:      1376
1571919693:      1466
1571919707:      1194
1571919719:      1166
1571919731:      1164
1571919744:      1164
1571919756:      1126
1571919768:      1126
1571919781:      1286
1571919795:      1188
1571919808:       994
1571919820:      1216
1571919834:      1178
1571919848:      1228
1571919861:      1156
1571919873:       962
1571919885:      1226
1571919899:      1256
1571919912:      1338
1571919926:      1312
1571919940:      1172
1571919953:      1154
1571919966:      1062
1571919977:       992
1571919989:      1356
1571920004:      1416
1571920020:      1412
1571920036:      1362
1571920050:      1254
1571920065:      1384
1571920079:      1208
1571920094:      1266
1571920108:      1176
1571920121:      1318
1571920134:      1306
1571920147:      1020
1571920158:      1038
1571920170:      1004
1571920180:      1064
1571920191:       980
1571920203:      1280
1571920218:      1464
1571920233:      1478
1571920247:      1234
1571920259:      1050
1571920273:      1304
1571920287:      1320
Count               Row Size        Cell Count
1                          0                 0
2                          0                 0
3                          0             93932
4                          0                 0
5                          0                 0
6                          0                 0
7                          0                 0
8                          0                 0
10                         0                 0
12                         0                 0
14                         0                 0
17                         0                 0
20                         0                 0
24                         0                 0
29                         0                 0
35                         0                 0
42                         0                 0
50                         0                 0
60                         0                 0
72                         0                 0
86                         0                 0
103                        0                 0
124                     4078                 0
149                    60488                 0
179                       91                 0
215                        0                 0
258                        0                 0
310                        0                 0
372                        0                 0
446                        0                 0
535                        0                 0
642                        0                 0
770                        0                 0
924                        0                 0
1109                       0                 0
1331                       0                 0
1597                       0                 0
1916                       0                 0
2299                       0                 0
2759                       0                 0
3311                       0                 0
3973                       0                 0
4768                      55                 0
5722                    3431                 0
6866                    8112                 0
8239                    7184                 0
9887                    4548                 0
11864                   2288                 0
14237                    987                 0
17084                   1307                 0
20501                    454                 0
24601                    334                 0
29521                    203                 0
35425                     77                 0
42510                     22                 0
51012                     14                 0
61214                     78                 0
73457                    177                 0
88148                      4                 0
105778                     0                 0
126934                     0                 0
152321                     0                 0
182785                     0                 0
219342                     0                 0
263210                     0                 0
315852                     0                 0
379022                     0                 0
454826                     0                 0
545791                     0                 0
654949                     0                 0
785939                     0                 0
943127                     0                 0
1131752                    0                 0
1358102                    0                 0
1629722                    0                 0
1955666                    0                 0
2346799                    0                 0
2816159                    0                 0
3379391                    0                 0
4055269                    0                 0
4866323                    0                 0
5839588                    0                 0
7007506                    0                 0
8409007                    0                 0
10090808                   0                 0
12108970                   0                 0
14530764                   0                 0
17436917                   0                 0
20924300                   0                 0
25109160                   0                 0
30130992                   0                 0
36157190                   0                 0
43388628                   0                 0
52066354                   0                 0
62479625                   0                 0
74975550                   0                 0
89970660                   0                 0
107964792                  0                 0
129557750                  0                 0
155469300                  0                 0
186563160                  0                 0
223875792                  0                 0
268650950                  0                 0
322381140                  0                 0
386857368                  0                 0
464228842                  0                 0
557074610                  0                 0
668489532                  0                 0
802187438                  0                 0
962624926                  0                 0
1155149911                 0                 0
1386179893                 0                 0
1663415872                 0                 0
1996099046                 0                 0
2395318855                 0                 0
2874382626                 0
3449259151                 0
4139110981                 0
4966933177                 0
5960319812                 0
7152383774                 0
8582860529                 0
10299432635                 0
12359319162                 0
14831182994                 0
17797419593                 0
21356903512                 0
25628284214                 0
30753941057                 0
36904729268                 0
44285675122                 0
53142810146                 0
63771372175                 0
76525646610                 0
91830775932                 0
110196931118                 0
132236317342                 0
158683580810                 0
190420296972                 0
228504356366                 0
274205227639                 0
329046273167                 0
394855527800                 0
473826633360                 0
568591960032                 0
682310352038                 0
818772422446                 0
982526906935                 0
1179032288322                 0
1414838745986                 0
Estimated cardinality: 73
```
I will try ` 'tombstone_threshold': '0.8'`.

**Comment by bancron** [comment]
Update:

I tried cherry-picking the [bug fix](https://github.com/scylladb/scylla/commit/dd769431259ddfea6b7523e45635e643fa11ef58) commit into the [3.1 release](https://github.com/scylladb/scylla/commit/9c3cdded9e2be1e4b4663b5465cc5259a3ad38a4) tag commit. There were merge conflicts so I wasn't able to try and build it.

I also changed some table compaction settings, which hasn't helped:

`compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '30', 'compaction_window_unit': 'MINUTES', 'tombstone_threshold': '0.8', 'unchecked_tombstone_compaction': 'true'}`

(previously `compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '1', 'compaction_window_unit': 'MINUTES'}`)

**Comment by slivne** [comment]
@bancron since you are not setting default time to live in the schema `default_time_to_live = 0` you need to set the TTL on insert/updates with `USING TTL X`

Lets try and validate this holds - can you try running sstabledump on one of the files and provide the output

as an example while I nserted a line to a sample table
```
insert into tribute.impressions (k, v) values (0x00,0x00) using ttl 10;
```

I got
```
[
  {
    "partition" : {
      "key" : [ "00" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 71,
        "liveness_info" : { "tstamp" : "2019-11-12T10:10:56.370491Z", "ttl" : 10, "expires_at" : "2019-11-12T10:11:06Z", "expired" : true },
        "cells" : [
          { "name" : "v", "value" : "0x00" }
        ]
      }
    ]
  }
]
```

that had TTL set.

**Comment by bancron** [comment]
Hi @slivne, 

I did insert every row with `USING TTL X`.

The issue was with using TimeWindowCompactionStrategy with multiple TTLs in a single table.

@avikivity [writes](https://scylladb-users.slack.com/archives/C2NLNBXLN/p1572357281282700), "Mixing TTLs doesn't work well with TWCS. So having a separate table for each TTL is really the only option."

I have switched to using a separate table per TTL and everything is working as expected now.
